lib/utils/utf8.c

```diff
@@ -79,3 +80,39 @@ char *utf8_lcpy(char *dst, const char *src, size_t n)
 
 	return dst;
 }
+
+bool utf8_check(const char *str, size_t maxlen)
+{
+	int i = 0, nbyte = 0;
+	size_t len = strlen(str);
+	unsigned char *buf = (unsigned char *)str;
+
+	/* It also return false when the string
+	 * is greater than the maximum length.
+	 */
+	if (len > maxlen) {
+		return false;
+	}
+
+	while (i < len) {
+		if (buf[i] <= 0x7F && buf[i] >= 0x00) {
+			i++;
+			continue;
+		} else {
+			if (buf[i] <= 0xDF && buf[i] >= 0xC2) {
+				nbyte = 2;
+			} else if (buf[i] <= 0xEF && buf[i] >= 0xE0) {
+				nbyte = 3;
+			} else if (buf[i] <= 0xF4 && buf[i] >= 0xF0) {
```

Comment:
Maybe you meant `0xF7` instead of `0xF4`?

It can help to pick the [edge cases](https://www.cl.cam.ac.uk/~mgk25/ucs/examples/UTF-8-test.txt) and see if they are all handled as expected.

For instance:
```
11100000 10000000 10000000          => decoded as \0
11100000 00000000 00000000          => invalid data
00000001                            => binary data
11110111 10000001 10111111 10111110 => https://symbl.cc/en/1FFE/
```

If it is a choice to not handle some cases, it would be good to have the list of what is not validated on the documentation.

Thanks! :)

Reply:
For the examples you provided, I will test them. However, I need to explain why it’s f0 and f4, not f0 and f7. f5 and f7 exceed the valid Unicode code points. While f0 to f7 appear to satisfy UTF-8, they are actually incorrect because UTF-8 is compatible with Unicode.

Reply:
Here are more examples in case you wished to make a more complete test suite:
```c
/* From https://www.cl.cam.ac.uk/~mgk25/ucs/examples/UTF-8-test.txt */

/* 1 Some correct UTF-8 text */

/* The Greek word 'kosme' */
zassert_...(utf8_check("\xce\xba\xe1\xbd\xb9\xcf\x83\xce\xbc\xce\xb5", 11));

/* 2.1 First possible sequence of a certain length

/* 2.1.1 1 byte  (U-00000000) */
zassert_...(utf8_check("\x00", 1));
/* 2.1.2 2 bytes (U-00000080) */
zassert_...(utf8_check("\xc2\x80", 2));
/* 2.1.3 3 bytes (U-00000800) */
zassert_...(utf8_check("\xe0\xa0\x80", 3));
/* 2.1.4 4 bytes (U-00010000) */
zassert_...(utf8_check("\xf0\x90\x80\x80", 4));
/* 2.1.5 5 bytes (U-00200000) */
zassert_...(utf8_check("\xf8\x88\x80\x80\x80", 5));
/* 2.1.6 6 bytes (U-04000000) */
zassert_...(utf8_check("\xfc\x84\x80\x80\x80\x80", 6));

/* 2.2 Last possible sequence of a certain length */

/* 2.2.1 1 byte  (U-0000007F) */
zassert_...(utf8_check("\x7f", 1));
/* 2.2.2 2 bytes (U-000007FF) */
zassert_...(utf8_check("\xdf\xbf", 2));
/* 2.2.3 3 bytes (U-0000FFFF) */
zassert_...(utf8_check("\xef\xbf\xbf", 3));
/* 2.2.4 4 bytes (U-001FFFFF) */
zassert_...(utf8_check("\xf7\xbf\xbf\xbf", 4));
/* 2.2.5 5 bytes (U-03FFFFFF) */
zassert_...(utf8_check("\xfb\xbf\xbf\xbf\xbf", 5));
/* 2.2.6 6 bytes (U-7FFFFFFF) */
zassert_...(utf8_check("\xfd\xbf\xbf\xbf\xbf\xbf", 6));

/* 2.3 Other boundary conditions */

/* 2.3.1 U-0000D7FF */
zassert_...(utf8_check("\xed\x9f\xbf", 3));
/* 2.3.2 U-0000E000 */
zassert_...(utf8_check("\xee\x80\x80", 3));
/* 2.3.3 U-0000FFFD */
zassert_...(utf8_check("\xef\xbf\xbd", 3));
/* 2.3.4 U-0010FFFF */
zassert_...(utf8_check("\xf4\x8f\xbf\xbf", 4));
/* 2.3.5 U-00110000 */
zassert_...(utf8_check("\xf4\x90\x80\x80", 4));

/* 3.1 Unexpected continuation bytes */

/* 3.1.1 First continuation byte 0x80 */
zassert_...(utf8_check("\x80", 1));
/* 3.1.2 Last continuation byte 0xbf */
zassert_...(utf8_check("\xbf", 1));
/* 3.1.3 2 continuation bytes */
zassert_...(utf8_check("\x80\xbf", 2));
/* 3.1.4 3 continuation bytes */
zassert_...(utf8_check("\x80\xbf\x80", 3));
/* 3.1.5 4 continuation bytes */
zassert_...(utf8_check("\x80\xbf\x80\xbf", 4));
/* 3.1.6 5 continuation bytes */
zassert_...(utf8_check("\x80\xbf\x80\xbf\x80", 5));
/* 3.1.7 6 continuation bytes */
zassert_...(utf8_check("\x80\xbf\x80\xbf\x80\xbf", 6));
/* 3.1.8 7 continuation bytes */
zassert_...(utf8_check("\x80\xbf\x80\xbf\x80\xbf\x80", 7));

/* 3.3 Sequences with last continuation byte missing */

/* 3.3.1 2-byte sequence with last byte missing (U+0000) */
zassert_...(utf8_check("\xc0", 1));
/* 3.3.2 3-byte sequence with last byte missing (U+0000) */
zassert_...(utf8_check("\xe0\x80", 2));
/* 3.3.3 4-byte sequence with last byte missing (U+0000) */
zassert_...(utf8_check("\xf0\x80\x80", 3));
/* 3.3.4 5-byte sequence with last byte missing (U+0000) */
zassert_...(utf8_check("\xf8\x80\x80\x80", 4));
/* 3.3.5 6-byte sequence with last byte missing (U+0000) */
zassert_...(utf8_check("\xfc\x80\x80\x80\x80", 5));
/* 3.3.6 2-byte sequence with last byte missing (U-000007FF) */
zassert_...(utf8_check("\xdf", 1));
/* 3.3.7 3-byte sequence with last byte missing (U-0000FFFF) */
zassert_...(utf8_check("\xef\xbf", 2));
/* 3.3.8 4-byte sequence with last byte missing (U-001FFFFF) */
zassert_...(utf8_check("\xf7\xbf\xbf", 3));
/* 3.3.9 5-byte sequence with last byte missing (U-03FFFFFF) */
zassert_...(utf8_check("\xfb\xbf\xbf\xbf", 4));
/* 3.3.10 6-byte sequence with last byte missing (U-7FFFFFFF) */
zassert_...(utf8_check("\xfd\xbf\xbf\xbf\xbf", 5));

/* 3.5 Impossible bytes */

/* 3.5.1 */
zassert_...(utf8_check("\xfe", 1));
/* 3.5.2 */
zassert_...(utf8_check("\xff", 1));
/* 3.5.3 */
zassert_...(utf8_check("\xfe\xfe\xff\xff", 4));

/* 4.1 Examples of an overlong ASCII character */

/* 4.1.1 U+002F */
zassert_...(utf8_check("\xc0\xaf", 2));
/* 4.1.2 U+002F */
zassert_...(utf8_check("\xe0\x80\xaf", 3));
/* 4.1.3 U+002F */
zassert_...(utf8_check("\xf0\x80\x80\xaf", 4));
/* 4.1.4 U+002F */
zassert_...(utf8_check("\xf8\x80\x80\x80\xaf", 5));
/* 4.1.5 U+002F */
zassert_...(utf8_check("\xfc\x80\x80\x80\x80\xaf", 6));

/* 4.2  Maximum overlong sequences */

/* 4.2.1 U-0000007F */
zassert_...(utf8_check("\xc1\xbf", 2));
/* 4.2.2 U-000007FF */
zassert_...(utf8_check("\xe0\x9f\xbf", 3));
/* 4.2.3 U-0000FFFF */
zassert_...(utf8_check("\xf0\x8f\xbf\xbf", 4));
/* 4.2.4 U-001FFFFF */
zassert_...(utf8_check("\xf8\x87\xbf\xbf\xbf", 5));
/* 4.2.5 U-03FFFFFF */
zassert_...(utf8_check("\xfc\x83\xbf\xbf\xbf\xbf", 6));

/* 4.3 Overlong representation of the NUL character */

/* 4.3.1 U+0000 */
zassert_...(utf8_check("\xc0\x80", 2));
/* 4.3.2 U+0000 */
zassert_...(utf8_check("\xe0\x80\x80", 3));
/* 4.3.3 U+0000 */
zassert_...(utf8_check("\xf0\x80\x80\x80", 4));
/* 4.3.4 U+0000 */
zassert_...(utf8_check("\xf8\x80\x80\x80\x80", 5));
/* 4.3.5 U+0000 */
zassert_...(utf8_check("\xfc\x80\x80\x80\x80\x80", 6));

/* 5.1 Single UTF-16 surrogates */

/* 5.1.1 U+D800 */
zassert_...(utf8_check("\xed\xa0\x80", 3));
/* 5.1.2 U+DB7F */
zassert_...(utf8_check("\xed\xad\xbf", 3));
/* 5.1.3 U+DB80 */
zassert_...(utf8_check("\xed\xae\x80", 3));
/* 5.1.4 U+DBFF */
zassert_...(utf8_check("\xed\xaf\xbf", 3));
/* 5.1.5 U+DC00 */
zassert_...(utf8_check("\xed\xb0\x80", 3));
/* 5.1.6 U+DF80 */
zassert_...(utf8_check("\xed\xbe\x80", 3));
/* 5.1.7 U+DFFF */
zassert_...(utf8_check("\xed\xbf\xbf", 3));

/* 5.2 Paired UTF-16 surrogates

/* 5.2.1 U+D800 U+DC00 */
zassert_...(utf8_check("\xed\xa0\x80\xed\xb0\x80", 6));
/* 5.2.2 U+D800 U+DFFF */
zassert_...(utf8_check("\xed\xa0\x80\xed\xbf\xbf", 6));
/* 5.2.3 U+DB7F U+DC00 */
zassert_...(utf8_check("\xed\xad\xbf\xed\xb0\x80", 6));
/* 5.2.4 U+DB7F U+DFFF */
zassert_...(utf8_check("\xed\xad\xbf\xed\xbf\xbf", 6));
/* 5.2.5 U+DB80 U+DC00 */
zassert_...(utf8_check("\xed\xae\x80\xed\xb0\x80", 6));
/* 5.2.6 U+DB80 U+DFFF */
zassert_...(utf8_check("\xed\xae\x80\xed\xbf\xbf", 6));
/* 5.2.7 U+DBFF U+DC00 */
zassert_...(utf8_check("\xed\xaf\xbf\xed\xb0\x80", 6));
/* 5.2.8 U+DBFF U+DFFF */
zassert_...(utf8_check("\xed\xaf\xbf\xed\xbf\xbf", 6));

/* 5.3 Noncharacter code positions */

/* 5.3.1 U+FFFE */
zassert_...(utf8_check("\xef\xbf\xbe", 3));
/* 5.3.2 U+FFFF */
zassert_...(utf8_check("\xef\xbf\xbf", 3));
```

Reply:
The `...` are there to let the author decide what it validates and what it does not validate.

Reply:
Emm, This test may be too long. Although it is very comprehensive.

Reply:
Yes UTF-8 + Unicode is complex :sweat_smile: I updated the `zassert_...` to be `zassert_true`/`zassert_false` so it's possible to check what works and what does not.

Maybe 4.1, 4.2 and 4.3 are the most important, as overlong sequences allow to inject `\0` bytes, and this can break some software that would blindly decode and re-encode the strings.

Reply:
> Emm, This test may be too long. Although it is very comprehensive.

Better to be thorough :)

Reply:
I decided to pick a representative one among many tests, because it makes almost the entire screen test.
(Honestly, there's no need.)

Reply:
And, I appreciate you finding this resource, and I will continue working on string encoding in the future!